### ISSUE TITLE
WebDAV REPORT sync-collection fails with empty sync-token

### ISSUE THREAD

**Issue by dilyanpalauzov** [cyrusimap/cyrus-imapd#4985]
DAVx⁵ 4.4 can do the communication below to synchronize collections.  Indeed, `report_sync_col()` contains: 
```c
r = sscanf((char *) str, SYNC_TOKEN_URL_SCHEME
  "%u-" MODSEQ_FMT "-" MODSEQ_FMT "%1s",
  &uidvalidity, &syncmodseq, &basemodseq,
  tokenuri /* test for trailing junk */);
if (r < 2 || r > 3 …) fctx->txn->error.desc = "Invalid sync-token";
else if (r == 3) {
  if (basemodseq > highestmodseq)
    fctx->txn->error.desc = "Invalid sync-token";
  else if (basemodseq < mailbox->i.deletedmodseq)
    fctx->txn->error.desc = "Stale sync-token";
} 
```
and I think the content of `str` is the content of &lt;sync-token /&gt; - so the empty string or NULL.

https://www.rfc-editor.org/rfc/rfc6578#section-3.4: 

> When the DAV:sync-collection request contains an empty DAV:sync-token element, the server MUST return all member URLs of the collection (taking account of the DAV:sync-level XML element value as per Section 3.3, and optional truncation of the result set as per Section 3.6) and it MUST NOT return any removed member URLs.  All types of member (collection or non-collection) MUST be reported.

For empty &lt;sync-token&gt; Cyrus IMAP returns “Invalid token” instead of all member URLs of the collection.
```
[sync.SyncManager] Starting initial sync
[sync.SyncManager] Number of local non-dirty entries: 0
[sync.SyncManager] Listing changes since null
--> REPORT https://cal.aegee.org/dav/calendars/user/cal/agenda/ h2
Depth: 0
User-Agent: DAVx5/4.4.1.1-ose (2024/07/13; dav4jvm; okhttp/4.12.0) Android/10
Accept-Language: bg-BG, bg;q=0.7, *;q=0.5
Content-Type: application/xml; charset=utf-8
Content-Length: 151
Host: cal.aegee.org
Connection: Keep-Alive
Accept-Encoding: gzip

<?xml version='1.0' encoding='UTF-8' ?><sync-collection xmlns="DAV:"><sync-token /><sync-level>1</sync-level><prop><getetag /></prop></sync-collection>
--> END REPORT (151-byte body)

[at.bitfire.dav4jvm.BasicDigestAuthHandler] Adding Basic authorization header for https://cal.aegee.org/dav/calendars/user/cal/agenda/
<-- 403 https://cal.aegee.org/dav/calendars/user/cal/agenda/ (322ms)
date: Mon, 22 Jul 2024 19:15:26 GMT
content-type: application/xml; charset=utf-8
content-length: 160
vary: Accept-Encoding

<?xml version="1.0" encoding="utf-8"?>
<D:error xmlns:D="DAV:"><D:valid-sync-token/><D:responsedescription>Invalid sync-token</D:responsedescription></D:error>

[network.HttpClient] <-- END HTTP (160-byte body)
[sync.SyncManager] Sync token invalid, performing initial sync
[sync.SyncManager] Number of local non-dirty entries: 0
 --> REPORT https://cal.aegee.org/dav/calendars/user/cal/agenda/ h2
 Depth: 0
 User-Agent: DAVx5/4.4.1.1-ose (2024/07/13; dav4jvm; okhttp/4.12.0) Android/10
 Accept-Language: bg-BG, bg;q=0.7, *;q=0.5
 Content-Type: application/xml; charset=utf-8
 Content-Length: 151
 Host: cal.aegee.org
 Connection: Keep-Alive
 Accept-Encoding: gzip
 
 <?xml version='1.0' encoding='UTF-8' ?><sync-collection xmlns="DAV:"><sync-token /><sync-level>1</sync-level><prop><getetag /></prop></sync-collection>
 --> END REPORT (151-byte body)
2024-07-22 22:15:27 17295 [at.bitfire.dav4jvm.BasicDigestAuthHandler] Adding Basic authorization header for https://cal.aegee.org/dav/calendars/user/cal/agenda/
<-- 403 https://cal.aegee.org/dav/calendars/user/cal/agenda/ (345ms)
date: Mon, 22 Jul 2024 19:15:27 GMT
content-type: application/xml; charset=utf-8
content-length: 160
vary: Accept-Encoding

<?xml version="1.0" encoding="utf-8"?>
<D:error xmlns:D="DAV:"><D:valid-sync-token/><D:responsedescription>Invalid sync-token</D:responsedescription></D:error>

[network.HttpClient] <-- END HTTP (160-byte body)
[sync.SyncManager] HTTP/DAV exception
EXCEPTION at.bitfire.dav4jvm.exception.ForbiddenException: HTTP 403 
	at at.bitfire.dav4jvm.DavResource.checkStatus(DavResource.kt:5)
	at at.bitfire.dav4jvm.DavResource.checkStatus(DavResource.kt:1)
	at at.bitfire.dav4jvm.DavResource.processMultiStatus(DavResource.kt:2)
	at at.bitfire.dav4jvm.DavCollection.reportChanges(DavCollection.kt:60)
	at at.bitfire.davdroid.sync.SyncManager.listRemoteChanges(SyncManager.kt:50)
	at at.bitfire.davdroid.sync.SyncManager.performSync$lambda$3$lambda$2(SyncManager.kt:99)
	at at.bitfire.davdroid.sync.SyncManager.$r8$lambda$eO8UdsZKI0OT7nVve-FhW6Qwd8Q(SyncManager.kt:1)
```

**Comment by ksmurchison** [comment]
What version of Cyrus is this issue occurring on?  Current Cyrus works fine:

`<<<<<<<< REPORT http://127.0.0.1:9101/dav/calendars/user/cassandane/72879F0E-58C3-11EF-A412-BB19DA390A17 HTTP/1.1
<sync-collection xmlns="DAV:">
  <sync-token />
  <sync-level>1</sync-level>
  <prop>
    <getetag />
  </prop>
</sync-collection>

>>>>>>>> HTTP/1.1 207 Multi-Status
<?xml version="1.0" encoding="utf-8"?>
<multistatus xmlns="DAV:">
  <response>
    <href>/dav/calendars/user/cassandane/72879F0E-58C3-11EF-A412-BB19DA390A17/851e34f4-23fc-4b69-9e90-67468336e53c.ics</href>
...
`

**Comment by ksmurchison** [comment]
What version of Cyrus is this occurring on?  Current Cyrus is fine.  I just wrote a Cassandane test:

`<<<<<<<< REPORT http://127.0.0.1:9101/dav/calendars/user/cassandane/72879F0E-58C3-11EF-A412-BB19DA390A17 HTTP/1.1
<sync-collection xmlns="DAV:">
  <sync-token />
  <sync-level>1</sync-level>
  <prop>
    <getetag />
  </prop>
</sync-collection>

>>>>>>>> HTTP/1.1 207 Multi-Status
<?xml version="1.0" encoding="utf-8"?>
<multistatus xmlns="DAV:">
  <response>
    <href>/dav/calendars/user/cassandane/72879F0E-58C3-11EF-A412-BB19DA390A17/851e34f4-23fc-4b69-9e90-67468336e53c.ics</href>
`

**Comment by dilyanpalauzov** [comment]
It is occurring on my Cyrus 3.8.  I do not know why the synchronization with DAVx⁵ has ever worked before —  https://github.com/bitfireAT/davx5-ose/discussions/952 . At https://github.com/cyrusimap/cyrus-imapd/pull/4992 I proposed a fix.

When you upload `<sync-token />` does `(str = xmlNodeListGetString(inroot->doc, node->children, 1))` on line 7624 for you return NULL or ""? For me it returns "".  https://gitlab.gnome.org/GNOME/libxml2/-/releases/v2.13.0 said “Rework xmlNodeListGetString” two months ago.

**Comment by ksmurchison** [comment]
It returns NULL for me.  libXML v2.13.1

**Comment by dilyanpalauzov** [comment]
```
curl -XREPORT -HDepth:0 -HAccept:application/xml -Hcontent-type:application/xml --data-binary @- <<EOF https://cal.aegee.org/dav/calendars/user/cal/agenda/

<sync-collection xmlns="DAV:"><sync-token /><sync-level>1</sync-level><prop><getetag /></prop></sync-collection>
EOF
```
Then after
```c
            if (!xmlStrcmp(node->name, BAD_CAST "sync-token") &&
                (str = xmlNodeListGetString(inroot->doc, node->children, 1))) {
```
`str` is `""` with libxml2 2.13.1.

**Comment by dilyanpalauzov** [comment]
https://gitlab.gnome.org/GNOME/libxml2/-/issues/783 — Why for empty XML elements xmlNodeListGetString() returns on one system NULL and on another "", both using 2.13.1.

**Comment by dilyanpalauzov** [comment]
So, at https://gitlab.gnome.org/GNOME/libxml2/-/issues/783 is stated that for empty XML elements xmlNodeListGetString() since 2.13.0 should return "", but this is a regression, and the next release of libxml2 will return again `NULL`.

**Comment by dilyanpalauzov** [comment]
libxml2 regression in 2.13.0, to be fixed in 2.13.4.
